RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco2D.cc

```diff
@@ -0,0 +1,654 @@
+      }
+   }
+   
+   ipass = 0;
+   if(npass == 1) {
+      // one pass, require that it have iterated
+      if(niter0[0] == 0) {return 2;}
+   } else {
+      // two passes
+      if(niter0[0] == 0 && niter0[1] == 0) {return 2;}
+      if(niter0[0] > 0 && niter0[1] > 0) {
+         // if both have iterated, take the smaller chi2
+         if(chi2min[1] < chi2min[0]) {ipass = 1;}
+      } else {
+         // if one has iterated, take it
+         if(niter0[1] > 0) {ipass = 1;}
+      }
+   }
+   
+   //   printf("niter = %d \n", niter);
+   
+   // Correct the charge ratio for missing pixels
+   
+   float fq;
+   if(qtfrac0[ipass] < 0.10f || qtfrac0[ipass] > 1.f) {qtfrac0[ipass] = 1.f;}
+   fq = fq0/qtfrac0[ipass];
+   
+   //   printf("qtfrac0 = %f \n", qtfrac0);
+   
+   if(fq > fbin[0]) {
+      qbin=0;
+   } else {
+      if(fq > fbin[1]) {
+         qbin=1;
+      } else {
+         if(fq > fbin[2]) {
+            qbin=2;
+         } else {
+            qbin=3;
+         }
+      }
+   }
+   
+   // Get charge related quantities
+   
+   float scalex = templ2D.scalex(qbin);
+   float scaley = templ2D.scaley(qbin);
+   float offsetx = templ2D.offsetx(qbin);
+   float offsety = templ2D.offsety(qbin);
+//   printf("scalex = %f, scaley = %f, offsetx = %f, offsety = %f \n", scalex, scaley, offsetx, offsety);
+   
+   // This 2D code has the origin (0,0) at the lower left edge of the input cluster
+   // That is now pixel [1,1] and the template reco convention is the middle
+   // of that pixel, so we need to correct
+   
+   xrec = x2D0[ipass] - xsize/2. - offsetx;
+   if(xdouble[0]) xrec -= xsize/2.f;
+   yrec = y2D0[ipass] - ysize/2. - offsety;
+   if(ydouble[0]) yrec -= ysize/2.f;
+   if(xerr2[ipass] > 0.f) {
+      sigmax = scalex*sqrt(xerr2[ipass]);
+      if(sigmax < 3.f) sigmax = 3.f;
+   } else {
+      sigmax = 10000.f;
+   }
+   if(yerr2[ipass] > 0.f) {
+      sigmay = scaley*sqrt(yerr2[ipass]);
+      if(sigmay < 3.f) sigmay = 3.f;
+   } else {
+      sigmay = 10000.f;
+   }
+   if(id > 0) {
+      // Do chi2 probability calculation
+      double meanxy = (double)(npixel*templ2D.chi2ppix());
+      double chi2scale = (double)templ2D.chi2scale();
+      if(meanxy < 0.01) {meanxy = 0.01;}
+      double hndof = meanxy/2.f;
+      double hchi2 = chi2scale*chi2min[ipass]/2.f;
+      probxy = (float)(1. - TMath::Gamma(hndof, hchi2));
+      // Do the charge probability
+      float mpv = templ2D.mpvvav();
+      float sigmaQ = templ2D.sigmavav();
+      float kappa = templ2D.kappavav();
+      float xvav = (qtotal/qtfrac0[ipass]-mpv)/sigmaQ;
+      float beta2 = 1.f;
+      //  VVIObj is a private port of CERNLIB VVIDIS
+      VVIObjF vvidist(kappa, beta2, 1);
+      float prvav = vvidist.fcn(xvav);
+      probQ = 1.f - prvav;
+   } else {
+      probxy = chi2min[ipass];
+      npixels = npixel;
+      probQ = 0.f;
+   }
+   
+   // Now
+   
+   //   printf("end minimization, errors = %f, %f \n\n", sqrt(xerr2), sqrt(yerr2));

```

Comment:
Are these commented out lines needed? If so, please add another line of comment which explain why; otherwise just remove them

Reply:
I don't know.  I'll need to discuss with Morris.  I propose to postpone this until later.